working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,948 @@
+<code>E<T<sub>1</sub>, .. T<sub>k</sub>></code>.
+
+During static analysis of the body of an extension type constructor, the
+return type is considered to be the on-type of the enclosing extension type
+declaration.
+
+It is a compile-time error if it is possible to reach the end of an
+extension type constructor without returning anything. *Even in the case
+where the on-type is nullable and the intended representation is the null
+object, an explicit `return null;` is required.*
+
+Let `E` be an explicit extension type declaration. It is not an error to
+declare a member in `E` which is also a member of `Object?`.
+
+*This differs from a non-explicit extension type declaration (which
+includes all existing declarations of extension methods), where such a
+member is a compile-time error. The rationale is that an extension method
+like `toString()` on an extension type `E` can be invoked on a receiver
+whose static type is `E`. It should be noted that the extension method
+named `toString` will only be invoked when the static receiver type is `E`,
+and, e.g., `(o as Object?).toString()` will invoke the implementation of
+`toString()` in the run-time class of `o`.*
+
+*The members of `Object?` are part of the interface of every extension
+type, and this means that it is a compile-time error if said member
+declaration is not a correct override of the declaration in `Object?`. For
+instance, `bool toString(int arg) => true;` is an error.*
+
+*It may not be obvious why we would want to prevent implicit invocations of
+the members of any given extension type. Here is a rationale:*
+
+*Consider the type `int`. This type is likely to be used as the on-type of
+many different extension types, because it allows a very lightweight object
+to play the role as a value with a specific interpretation (say, an `Age`
+in years or a `Width` in pixels). Different extension types are not
+assignable to each other, so we'll offer a certain protection against
+inconsistent interpretations.*
+
+*However, if we have many different extension types with the same or
+overlapping on-types, then it may be impractical to work with: Lots of
+extension methods are applicable to any given expression of that on-type,
+and they are not intended to be used at all, each of them should only be
+used when the associated interpretation is valid, that is, when the static
+type of the receiver is the extension type that declares said member.*
+
+*Hence, we want to support the notion of an extension type whose methods
+are never invoked implicitly. An explicit extension type will do this.*
+
+
+### Allow instance member access using `show` and `hide`
+
+This section specifies the effect of including a non-empty
+`<extensionShowHidePart>` in an extension declaration.
+
+*The show/hide part provides access to a subset of the members of the
+interface of the on-type. For instance, there may be some read-only methods
```

Comment:
Why is "read-only" relevant?  There might be "write-only" methods that we can safely call as well, right?

Reply:
Sure, but the case where we wish to maintain an invariant which is specified in terms of the state of the object which is denoted by `this`, it's obvious that the invariant won't be violated by calling a method that doesn't change that state (so a read-only method is a semantic concept, it also can't call any other code that performs any such modification, also not when it happens via an alias to `this`).

A write-only method would be known to preserve an invariant in a much more marginal set of cases: For instance, the invariant could be associated with parts of the state not including the parts that this method would change (for instance, when we are dealing with a "semantically immutable" object which uses caching to avoid repeating certain computations).

At this point I just want to give an example, so I preferred mentioning the obvious case and ignoring the tricky marginal case.

Reply:
Up to you, but I found the reference to read-only just distracting.  

> t's obvious that the invariant won't be violated by calling a method that doesn't change that state

This isn't obvious to me.  I can build lots of abstractions that rely on certain properties of an object not being read.  I mean, abstract types are thing in purely functional languages without state at all!